Initial `java_fuzz_test` implementation

fmeum opened the pull request:
Adds a java_fuzz_test macro for Java fuzz tests that can be invoked with the Java fuzzer Jazzer.

Most of the existing code remains unchanged, with the following exceptions:

* `CcFuzzingEngine`, `CcFuzzingBinaryInfo`, `cc_fuzzing_engine` and `CcFuzzingEngineInfo` lose their `cc` prefix as they can also be used or augmented for other languages, including Java.
* `CcFuzzingEngineInfo` and `fuzzing_engine` receive an additional `java_library(_info)` attribute/field that contains an optional Java   library target that all Java fuzz targets should depend on.

A small architectural difference between Jazzer and e.g. libFuzzer fuzz targets is that the fuzzer itself is not linked into Java fuzz targets, but remains a separate, stand-alone binary (the "driver"). The `java_fuzz_test` macro deals with this by applying the instrumentation transition to native library dependencies directly and creating a fuzzing_binary_uninstrumented based on a binary shim that launches the driver.

@stefanbucur I have left a few FIXME comments at points where I ran into Bazel issues I couldn't solve. If you have an idea how to resolve any of the issues explained there, I would be very grateful. The PR is however fully functional (apart from the OSS-Fuzz builds), so please give it a try and let me know what you think.

Missing:

- [ ] OSS-Fuzz support
- [ ] User-facing documentation

Fixes #141.

fmeum commented on fuzzing/private/java.bzl:
```diff
@@ -0,0 +1,168 @@
+            # Build in target configuration rather than host because the driver
+            # uses transitions to set the correct C++ standard for its
+            # dependencies.
+            cfg = "target",
```
The `jazzer_driver` target uses a transition to apply the copt`-std=c++17` to it and all dependencies. Without this, it seems that any workspace depending on `rules_fuzzing` would need to set this in their `.bazelrc`. If you know a better way to build the driver with C++17 without affecting the rest of the workspace, please let me know.

stefanbucur commented:
Thanks a lot for this! Apologies for the delay, I've been swamped with work today, but will get back to review this PR as soon as possible.

fmeum commented:
I pushed a fix for the `cpp_lint` failure as a separate commit (feel free to squash it).

The regression test probably also needs some changes as it currently tries to run Java targets with a honggfuzz config. Do you have a suggestion on what to test here with the existing Java targets? I could also integrate a more interesting Java fuzzer that finds a real bug, e.g. based on an OSS-Fuzz project.

stefanbucur commented on examples/BUILD:
```diff
@@ -111,3 +112,20 @@ cc_fuzz_test(
 )
+
+java_fuzz_test(
+    name = "JavaFuzzTest",
```
We use lower_case naming conventions for Bazel targets. Should be `java_fuzz_test`.

Also by convention we alphabetize the target names in the BUILD file.

fmeum replied:
> We use lower_case naming conventions for Bazel targets. Should be `java_fuzz_test`.

Instances of `java_binary` and `java_test` are customarily named after the main class, which allows for omitting the `main_class` attribute (`java_library` targets use lower_ case, which lines up with package names in Java). I felt that it would be useful for `java_fuzz_test` to follow the same convention with respect to it's `target_class`. See https://docs.bazel.build/versions/master/be/java.html#java_binary_args for more details.

> Also by convention we alphabetize the target names in the BUILD file.

Will do.

stefanbucur replied:
> > We use lower_case naming conventions for Bazel targets. Should be `java_fuzz_test`.
> 
> Instances of `java_binary` and `java_test` are customarily named after the main class, which allows for omitting the `main_class` attribute (`java_library` targets use lower_ case, which lines up with package names in Java). I felt that it would be useful for `java_fuzz_test` to follow the same convention with respect to it's `target_class`. See https://docs.bazel.build/versions/master/be/java.html#java_binary_args for more details.

Ah! I understand now, thank you for the extra context.

stefanbucur commented on examples/BUILD:
```diff
@@ -111,3 +112,20 @@ cc_fuzz_test(
+)
+
+java_fuzz_test(
+    name = "JavaNativeFuzzTest",
```
Ditto here: `java_native_fuzz_test`.

fmeum replied:
See my comment above.

stefanbucur commented on examples/src/main/java/com/example/JavaFuzzTest.java:
```diff
@@ -0,0 +1,24 @@
+package com.example;
```
What is the role of the `src/main/java/com/example` directory structure? Is it mandated by Bazel? (I'm not quite familiar with Java rules.)

If this is mandatory, I was going to suggest to place the Java code under an `examples/java` directory. I.e., this would become `examples/java/src/...`. This way, it will be clear why some code requires this nesting structure.

It might also make sense to place the Java fuzz targets under `examples/java`, but I'm fine either way.

fmeum replied:
While not required, it is certainly recommended: https://docs.bazel.build/versions/master/tutorial/java.html

I like the idea of moving the tests to a `java` subpackage. 

Should I leave the C targets as is or move them into a `cc` package?

stefanbucur replied:
> While not required, it is certainly recommended: https://docs.bazel.build/versions/master/tutorial/java.html
> 
> I like the idea of moving the tests to a `java` subpackage.

Since we also need a different BUILD target naming convention, I'm becoming more in favor of this alternative too.

> 
> Should I leave the C targets as is or move them into a `cc` package?

If this was done from the beginning, that would have made sense, but now I feel it would be disruptive for anyone already using the library and expecting to find these examples here (e.g., in scripts). So we can leave them here for now and reconsider later if needed.

fmeum replied:
On second thought, moving the `java_fuzz_test`s into `java` directory messes up the `target_class` detection. Can I move them to `examples_java` instead, leaving the `cc_fuzz_test`s in `example`?

stefanbucur replied:
> On second thought, moving the `java_fuzz_test`s into `java` directory messes up the `target_class` detection. 

I'm not sure I follow this part. How does the base path of the BUILD file make a difference? I.e., does it matter the Java tree is rooted at `examples/{BUILD,src/main/...}` vs. `examples/java/{BUILD,src/main/...}`?

fmeum replied:
It does. The heuristic for determining the main class does scan the full path for the first occurence of `java` or `src` (see https://github.com/bazelbuild/bazel/blob/578a72c8448c21ecdb6ea7585b9da348fa1d7fd5/src/main/java/com/google/devtools/build/lib/rules/java/JavaCommon.java#L504).

We could of course just specify the `target_class` attribute in the examples, but that wouldn't tell users that they can expect the same semantics as from `java_binary`.

fmeum replied:
I found a way to please the heuristic and move everything into `examples/java`.

stefanbucur commented on examples/src/main/native/BUILD.bazel:
```diff
@@ -0,0 +1,48 @@
+load("@rules_cc//cc:defs.bzl", "cc_binary")
```
Rename `BUILD.bazel` to `BUILD`, for consistency?

stefanbucur commented on examples/src/main/java/com/example/JavaFuzzTest.java:
```diff
@@ -0,0 +1,24 @@
+package com.example;
```
Add copyright header?

stefanbucur commented on examples/src/main/java/com/example/JavaNativeFuzzTest.java:
```diff
@@ -0,0 +1,20 @@
+package com.example;
```
Add copyright header?

stefanbucur commented on examples/src/main/native/com_example_JavaNativeFuzzTest.h:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Use same comment style? (`//`)

stefanbucur commented on fuzzing/private/BUILD:
```diff
@@ -24,5 +24,21 @@ exports_files([
 ])
+
+# Config settings needed for java_fuzz_test.
+config_setting(
+    name = "use_jazzer_driver",
+    flag_values = {
+        "@rules_fuzzing//fuzzing:cc_engine_sanitizer": "none",
+    },
+)
```
Wouldn't this match for any driver with no sanitization enabled?

fmeum replied:
That is true. It's only referenced from the `java_fuzz_test` macro though. Would this be a problem? I could also match the `java_engine` here, but would then also need to set its default value to something other than `jazzer` for this to no longer match by default.

stefanbucur replied:
I think it is fine to match the sanitizer flag only - but in this case maybe the name of the config setting should be more specific (basically, name it in the spirit of what it does vs. how it is being used). For example, "use_no_sanitizer", or "engine_sanitizer_none"?

stefanbucur commented on fuzzing/repositories.bzl:
```diff
@@ -18,7 +18,7 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+def rules_fuzzing_dependencies(oss_fuzz = True, honggfuzz = True, jazzer = True):
     """Instantiates the dependencies of the fuzzing rules.
 
     Args:
```
Update the docstring, too?

stefanbucur commented on fuzzing/private/binary.bzl:
```diff
@@ -14,7 +14,7 @@
 """Defines a rule for creating an instrumented fuzzing executable."""
 
-load("//fuzzing/private:engine.bzl", "CcFuzzingEngineInfo")
+load("//fuzzing/private:engine.bzl", "FuzzingEngineInfo")
```
The `CcFuzzing*` -> `Fuzzing*` refactoring seems to impact quite a few files. Should we do it as a separate PR, to facilitate reviewing and keeping the changes more focused?

fmeum replied:
Yes, I'll split this up.

stefanbucur replied:
Maybe let's hold this off for a bit until we make a decision on the provider structure below.

stefanbucur commented on fuzzing/private/engine.bzl:
```diff
@@ -44,18 +45,33 @@ def _cc_fuzzing_engine_impl(ctx):
 
-    cc_library_info = ctx.attr.library[CcInfo]
-    cc_fuzzing_engine_info = CcFuzzingEngineInfo(
+    cc_library_info = None
+    java_library_info = None
+    default_info = None
+    providers = []
+    if ctx.attr.library and not ctx.attr.java_library:
+        cc_library_info = ctx.attr.library[CcInfo]
+        default_info = ctx.attr.library[DefaultInfo]
+        providers += [default_info, cc_library_info]
+    elif ctx.attr.java_library and not ctx.attr.library:
+        java_library_info = ctx.attr.java_library[JavaInfo]
+        default_info = ctx.attr.java_library[DefaultInfo]
+        providers += [default_info, java_library_info]
+    else:
+        fail("A fuzzing_engine must have exactly one of library and java_library set.")
+
+    fuzzing_engine_info = FuzzingEngineInfo(
```
Design question: Are we sharing the same provider here in order to reuse the shared logic of populating these fields?

I'm wondering if we could still take advantage of the Bazel provider type system and define a `CommonFuzzingEngineInfo` provider, that can be nested inside separate `CcFuzzingEngineInfo` and `JavaFuzzingEngineInfo` providers.

fmeum replied:
If it's possible to nest providers than this would certainly be the preferable solution. Do you know whether that is possible? The documentation at https://docs.bazel.build/versions/master/skylark/lib/Provider.html suggests it may not be.

stefanbucur replied:
That should be possible - we are actually already nesting providers this way (e.g., the `FuzzingEngienInfo.java_library_info` field includes a `JavaInfo` provider.

I'm basically thinking of the following setup:

```
CommonFuzzingEngineInfo = provider(
   ...
)

CcFuzzingEngineInfo = provider(
   cc_library_info = ...
   engine_info = (the CommonFuzzingEngineInfo data)
)

JavaFuzzingEngineInfo = provider(
    java_library_info = ...
    engine_info = (CommonFuzzingEngineInfo data)
)
```

We can then have two distinct rules java_fuzzing_engine and cc_fuzzing_engine instantiating these providers. We can factor out shared functionality (the dict of shared attributes, a shared function to compute CommonFuzzingEngineInfo).

WDYT?

fmeum replied:
How about we drop the `*_library_info` field entirely in favor of a single, language-independent `FuzzingEngineInfo` provider? Engines already expose their libraries via a dedicated `{Cc, Java}Info` provider and the single use of that info in the code base is via that provider anyway. 

Composing providers rather than nesting them also interacts better with the `provides` directive of the `engine` attribute in `fuzzing_binary`. So even if we want to keep language-specific engine info providers, it may be easier to split them into language-dependent and language-independent parts rather than nesting them. What do you think?

stefanbucur replied:
Do you mean to basically say that attributes depending on a fuzzing engine should explicitly list the extra `{Cc,Java}Info` providers they might need? E.g.:

```
"engine": attr.label(providers=[FuzzingEngineInfo,CcInfo]),
```

I think that's a very good idea and a better use of Bazel providers than what we currently do!

Would it make sense to do this first in a separate PR, just so we don't mix semantics-preserving refactorings with adding new functionality?

fmeum replied:
I submitted #144 with just this change. As it turns out, there are no attributes that explicitly depend on the language info provider, which makes the refactoring even easier.

stefanbucur reviewed (changes requested):
I did a first scan of the code and left a number of initial comments.

stefanbucur commented on fuzzing/private/BUILD:
```diff
@@ -24,5 +24,21 @@ exports_files([
+)
+
+config_setting(
+    name = "use_jazzer_driver_asan",
```
Ditto here: "use_asan_sanitizer" or "engine_sanitizer_asan"?

fmeum commented:
I have addressed the comments and rebased this PR onto #144, which should be reviewed first.

Along the way, I resolved one of the FIXMEs by reading the transitive native library info directly from the `JavaInfo` provider if supported by the current Bazel version (likely to land in Bazel 5).

fmeum commented:
I rebased the PR onto master, it should be ready for review now.

stefanbucur commented on examples/java/com/example/JavaNativeFuzzTest.h:
```diff
@@ -0,0 +1,35 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* DO NOT EDIT THIS FILE - it is machine generated */
```
Is this generated as part of the build? If yes, do we need to check it into the repository?

fmeum replied:
It was generated by a JDK, but there is no need to preserve that information in a comment. I have removed it.

stefanbucur commented on fuzzing/engines/jazzer_launcher.sh:
```diff
@@ -0,0 +1,42 @@
+# Copyright 2020 Google LLC
```
2021 for new files (here and everywhere else it applies).

fmeum replied:
> The CI checks also need to be updated, right? I think we should have separate matrices for C++ (libFuzzer / Honggfuzz / etc.) and Java tests (Jazzer / any other engines?).
> 
> Otherwise, the general architecture looks good to me.

Thanks! I create one more Java target that keeps running and added the Java tests as separate matrix tests to the workflow file.

stefanbucur commented on fuzzing/engines/jazzer_launcher.sh:
```diff
@@ -0,0 +1,42 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# The command line arguments for launching fuzz tests run with the libFuzzer
```
I see the comment mentions libFuzzer. In a previous revision, you mentioned the launcher interface is the same. Should we reuse the libFuzzer one instead of duplicating it here?

fmeum replied:
Jazzer is compatible with libFuzzer in the sense that it understands all libFuzzer arguments. It does take additional (double-dash) arguments though and I initially forgot that we need to pass it `--reproducer_path` so that Java reproducers are also stored in the artifact dir.

I restructured this file to make this clearer.

stefanbucur commented on fuzzing/repositories.bzl:
```diff
@@ -18,12 +18,15 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 load("//fuzzing/private/oss_fuzz:repository.bzl", "oss_fuzz_repository")
 
-def rules_fuzzing_dependencies(oss_fuzz = True, honggfuzz = True):
+def rules_fuzzing_dependencies(oss_fuzz = True, honggfuzz = True, jazzer = True):
```
Existing (C++) users of this library upgrading to this revision will now automatically pull the Jazzer dependencies in their workspace. Since this is adding an unused dep to all existing users, how about turning it off by default? Depending on future usage information, we can decide to turn it back on.

fmeum replied:
Sounds good.

stefanbucur commented on java_examples/src/main/native/BUILD:
```diff
@@ -0,0 +1,48 @@
+load("@rules_cc//cc:defs.bzl", "cc_binary")
```
Is this file still needed?

fmeum replied:
No, thanks, I deleted it.

stefanbucur reviewed (changes requested):
The CI checks also need to be updated, right? I think we should have separate matrices for C++ (libFuzzer / Honggfuzz / etc.) and Java tests (Jazzer / any other engines?).

Otherwise, the general architecture looks good to me.

fmeum commented:
> The CI checks also need to be updated, right? I think we should have separate matrices for C++ (libFuzzer / Honggfuzz / etc.) and Java tests (Jazzer / any other engines?).
> 
> Otherwise, the general architecture looks good to me.

Thanks! I create one more Java target that keeps running and added the Java tests as separate matrix tests to the workflow file.

stefanbucur commented on fuzzing/private/java_utils.bzl:
```diff
@@ -0,0 +1,192 @@
+    "sanitizer_configs",
+)
+
+"""Utilities and helper rules for Java fuzz tests."""
```
Nit: Move before the load() statements.

stefanbucur reviewed (approved):
This looks great!

fmeum commented:
Thanks! I added a commit that fixes the nit as well as the `//docs:cc_rules_fuzzing` build. The docs for Java fuzz tests can come in a follow-up PR.

I haven't squashed the commits yet, but feel free to do so.
